File: java/test/integration/core/ExampleTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.typedb.driver.test.integration.core;

// EXAMPLE START MARKER
import com.typedb.driver.TypeDB;
import com.typedb.driver.api.Driver;
import com.typedb.driver.api.QueryType;
import com.typedb.driver.api.Transaction;
import com.typedb.driver.api.answer.ConceptRow;
import com.typedb.driver.api.answer.ConceptRowIterator;
import com.typedb.driver.api.answer.QueryAnswer;
import com.typedb.driver.api.concept.Concept;
import com.typedb.driver.api.concept.type.AttributeType;
import com.typedb.driver.api.concept.type.EntityType;
import com.typedb.driver.api.database.Database;
import com.typedb.driver.common.Promise;
import com.typedb.driver.common.exception.TypeDBDriverException;
// EXAMPLE END MARKER

import org.junit.BeforeClass;
import org.junit.Test;

// EXAMPLE START MARKER
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

// EXAMPLE END MARKER

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

@SuppressWarnings("Duplicates")
// EXAMPLE START MARKER
public class ExampleTest {
    // EXAMPLE END MARKER
    @BeforeClass
    public static void setUpClass() {
        Driver typedbDriver = TypeDB.coreDriver(TypeDB.DEFAULT_ADDRESS);
        if (typedbDriver.databases().contains("typedb")) {
            typedbDriver.databases().get("typedb").delete();
        }
        typedbDriver.close();
    }

    @Test
    // EXAMPLE START MARKER
    public void example() {
        // Open a driver connection. Try-with-resources can be used for automatic driver connection management
        try (Driver driver = TypeDB.coreDriver(TypeDB.DEFAULT_ADDRESS)) {
            // Create a database
            driver.databases().create("typedb");
            Database database = driver.databases().get("typedb");
            assertEquals(database.name(), "typedb");

            // Open transactions of 3 types
            Transaction tx = driver.transaction(database.name(), Transaction.Type.READ);

            // Use "try" blocks to catch driver exceptions
            try {
                // Execute any TypeDB query using TypeQL. Wrong queries are rejected with an explicit exception
                Promise<? extends QueryAnswer> promise = tx.query("define entity i-cannot-be-defined-in-read-transactions;");

                System.out.println("The result is still promised, so it needs resolving even in case of errors!");
                promise.resolve();
            } catch (TypeDBDriverException expectedException) {
                System.out.println("Once the query's promise is resolved, the exception is revealed: " + expectedException);
            } finally {
                // Don't forget to close the transaction!
                tx.close();
            }

            // Open a schema transaction to make schema changes
            // Use try-with-resources blocks to forget about "close" operations (similarly to connections)
            try (Transaction transaction = driver.transaction(database.name(), Transaction.Type.SCHEMA)) {
                String defineQuery = "define " +
                        "entity person, owns name, owns age; " +
                        "attribute name, value string;\n" +
                        "attribute age, value long;";

                QueryAnswer answer = transaction.query(defineQuery).resolve();
                assertTrue(answer.isOk());
                assertEquals(QueryType.SCHEMA, answer.getQueryType());

                // Commit automatically closes the transaction. It can still be safely called inside "try" blocks
                transaction.commit();
            }


            // Open a read transaction to safely read anything without database modifications
            try (Transaction transaction = driver.transaction(database.name(), Transaction.Type.READ)) {
                QueryAnswer entityAnswer = transaction.query("match entity $x;").resolve();
                assertTrue(entityAnswer.isConceptRows());
                assertFalse(entityAnswer.isConceptDocuments());
                assertEquals(QueryType.READ, entityAnswer.getQueryType());

                // Collect concept rows that represent the answer as a table
                List<ConceptRow> entityRows = entityAnswer.asConceptRows().stream().collect(Collectors.toList());
                assertEquals(entityRows.size(), 1);
                ConceptRow entityRow = entityRows.get(0);

                // Collect column names to get concepts by index if the variable names are lost
                List<String> entityHeader = entityRow.columnNames().collect(Collectors.toList());
                assertEquals(entityHeader.size(), 1);

                String columnName = entityHeader.get(0);
                assertEquals(columnName, "x");

                // Get concept by the variable name (column name)
                Concept conceptByName = entityRow.get(columnName);

                // Get concept by the header's index
                Concept conceptByIndex = entityRow.getIndex(0);
                assertEquals(conceptByName, conceptByIndex);

                // Check if it's an entity type before the conversion
                if (conceptByName.isEntityType()) {
                    System.out.printf("Getting concepts by variable names and indexes is equally correct. " +
                                    "Both represent the defined entity type: '%s' (in case of a doubt: '%s')%n",
                            conceptByName.asEntityType().getLabel(),
                            conceptByIndex.asEntityType().getLabel());
                }
                assertTrue(conceptByName.isEntityType());
                assertTrue(conceptByName.isType());
                assertEquals(conceptByName.asEntityType().getLabel(), "person");
                assertNotEquals(conceptByName.asEntityType().getLabel(), "not person");
                assertNotEquals(conceptByName.asEntityType().getLabel(), "age");

                // Continue querying in the same transaction if needed
                QueryAnswer attributeAnswer = transaction.query("match attribute $a;").resolve();
                assertTrue(attributeAnswer.isConceptRows());
                assertEquals(QueryType.READ, attributeAnswer.getQueryType());

                // ConceptRowIterator can be used as any other Iterator
                ConceptRowIterator attributeRowIterator = attributeAnswer.asConceptRows();

                while (attributeRowIterator.hasNext()) {
                    ConceptRow attributeRow = attributeRowIterator.next();

                    // Column names are a stream, so they can be used in a similar way
                    Iterator<String> columnNameIterator = attributeRow.columnNames().iterator();
                    columnName = columnNameIterator.next();
                    assertFalse(columnNameIterator.hasNext());

                    conceptByName = attributeRow.get(columnName);

                    // Check if it's an attribute type before the conversion
                    if (conceptByName.isAttributeType()) {
                        AttributeType attributeType = conceptByName.asAttributeType();
                        System.out.printf("Defined attribute type's label: '%s', value type: '%s'%n", attributeType.getLabel(), attributeType.getValueType());
                        assertTrue(attributeType.isLong() || attributeType.isString());
                        assertTrue(Objects.equals(attributeType.getValueType(), "long") || Objects.equals(attributeType.getValueType(), "string"));
                        assertTrue(Objects.equals(attributeType.getLabel(), "age") || Objects.equals(attributeType.getLabel(), "name"));
                        assertNotEquals(attributeType.getLabel(), "person");
                        assertNotEquals(attributeType.getLabel(), "person:age");

                        System.out.printf("It is also possible to just print the concept itself: '%s'%n", conceptByName);
                        assertTrue(conceptByName.isAttributeType());
                        assertTrue(conceptByName.isType());
                    }
                }
            }

            // Open a write transaction to insert data
            try (Transaction transaction = driver.transaction(database.name(), Transaction.Type.WRITE)) {
                String insertQuery = "insert $z isa person, has age 10; $x isa person, has age 20, has name \"John\";";
                QueryAnswer answer = transaction.query(insertQuery).resolve();
                assertTrue(answer.isConceptRows());
                assertEquals(QueryType.WRITE, answer.getQueryType());

                // Insert queries also return concept rows
                List<ConceptRow> rows = answer.asConceptRows().stream().collect(Collectors.toList());
                assertEquals(rows.size(), 1);
                ConceptRow row = rows.get(0);
                row.columnNames().iterator().forEachRemaining(columnName -> {
                    Concept insertedConcept = row.get(columnName);
                    System.out.printf("Successfully inserted $%s: %s%n", columnName, insertedConcept);
                    if (insertedConcept.isEntity()) {
                        System.out.println("This time, it's an entity, not a type!");
                    }
                });

                // It is possible to ask for the column names again
                List<String> header = row.columnNames().collect(Collectors.toList());
                assertEquals(header.size(), 2);
                assertTrue(header.contains("x"));
                assertTrue(header.contains("z"));

                Concept x = row.getIndex(header.indexOf("x"));
                if (x.isEntity()) {
                    System.out.println("Each entity receives a unique IID. It can be retrieved directly: " + x.asEntity().getIID());
                }

                // Do not forget to commit if the changes should be persisted
                transaction.commit();
            }

            // Open another write transaction to try inserting even more data
            try (Transaction transaction = driver.transaction(database.name(), Transaction.Type.WRITE)) {
                // When loading a large dataset, it's often better not to resolve every query's promise immediately.
                // Instead, collect promises and handle them later. Alternatively, if a commit is expected in the end,
                // just call `commit`, which will wait for all ongoing operations to finish before executing.
                List<String> queries = List.of("insert $a isa person, has name \"Alice\";", "insert $b isa person, has name \"Bob\";");
                for (String query : queries) {
                    transaction.query(query);
                }
                transaction.commit();
            }

            try (Transaction transaction = driver.transaction(database.name(), Transaction.Type.WRITE)) {
                // Commit will still fail if at least one of the queries produce an error.
                List<String> queries = List.of("insert $c isa not-person, has name \"Chris\";", "insert $d isa person, has name \"David\";");
                List<Promise<? extends QueryAnswer>> promises = new ArrayList<>();
                for (String query : queries) {
                    promises.add(transaction.query(query));
                }

                try {
                    transaction.commit();
                    fail("TypeDBDriverException is expected");
                } catch (TypeDBDriverException expectedException) {
                    System.out.println("Commit result will contain the unresolved query's error: " + expectedException);
                }
            }

            // Open a read transaction to verify that the inserted data is saved
            try (Transaction transaction = driver.transaction(database.name(), Transaction.Type.READ)) {
                // A match query can be used for concept row outputs
                String var = "x";
                QueryAnswer matchAnswer = transaction.query(String.format("match $%s isa person;", var)).resolve();
                assertTrue(matchAnswer.isConceptRows());
                assertEquals(QueryType.READ, matchAnswer.getQueryType());

                // Simple match queries always return concept rows
                AtomicInteger matchCount = new AtomicInteger(0);
                matchAnswer.asConceptRows().stream().forEach(row -> {
                    assertEquals(QueryType.READ, row.getQueryType());
                    Concept x = row.get(var);
                    assertTrue(x.isEntity());
                    assertFalse(x.isEntityType());
                    assertFalse(x.isAttribute());
                    assertFalse(x.isType());
                    assertTrue(x.isInstance());
                    EntityType xType = x.asEntity().getType().asEntityType();
                    assertEquals(xType.getLabel(), "person");
                    assertNotEquals(xType.getLabel(), "not person");
                    matchCount.incrementAndGet();
                    System.out.printf("Found a person %s of type %s%n", x, xType);
                });
                assertEquals(matchCount.get(), 4);
                System.out.println("Total persons found: " + matchCount.get());

                // A fetch query can be used for concept document outputs with flexible structure
                QueryAnswer fetchAnswer = transaction.query("match" +
                        "  $x isa! person, has $a;" +
                        "  $a isa! $t;" +
                        "fetch {" +
                        "  \"single attribute type\": $t," +
                        "  \"single attribute\": $a," +
                        "  \"all attributes\": { $x.* }," +
                        "};").resolve();
                assertTrue(fetchAnswer.isConceptDocuments());
                assertEquals(QueryType.READ, fetchAnswer.getQueryType());

                // Fetch queries always return concept documents
                AtomicInteger fetchCount = new AtomicInteger(0);
                fetchAnswer.asConceptDocuments().stream().forEach(document -> {
                    assertNotNull(document);
                    System.out.println("Fetched a document: " + document);
                    System.out.print("This document contains an attribute of type: ");
                    System.out.println(document.asObject().get("single attribute type").asObject().get("label"));

                    fetchCount.incrementAndGet();
                });
                assertEquals(fetchCount.get(), 5);
                System.out.println("Total documents fetched: " + fetchCount.get());
            }
        }

        System.out.println("More examples can be found in the API reference and the documentation.\nWelcome to TypeDB!");
    }
}
// EXAMPLE END MARKER

File: csharp/Test/Integration/ExampleTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// EXAMPLE START MARKER
using System;
using System.Collections.Generic;
using System.Linq;

using TypeDB.Driver;
using TypeDB.Driver.Api;
using TypeDB.Driver.Api.Answer;
using TypeDB.Driver.Common;

// EXAMPLE END MARKER
using NUnit.Framework;
// EXAMPLE START MARKER

namespace TypeDB.Driver.Test.Integration
{
    [TestFixture]
    public class ExampleTest
    {
        // EXAMPLE END MARKER
        [SetUp]
        public void SetUp()
        {
            // Clean up any existing test database before each test
            using var driver = TypeDB.Driver(TypeDB.DefaultAddress, new Credentials("admin", "password"), new DriverOptions(DriverTlsConfig.Disabled()));
            if (driver.Databases.Contains("typedb"))
            {
                driver.Databases.Get("typedb").Delete();
            }
        }

        [Test]
        // EXAMPLE START MARKER
        public void Example()
        {
            // Open a driver connection. Using statements can be used for automatic driver connection management
            using (var driver = TypeDB.Driver(TypeDB.DefaultAddress, new Credentials("admin", "password"), new DriverOptions(DriverTlsConfig.Disabled())))
            {
                // Create a database
                driver.Databases.Create("typedb");
                var database = driver.Databases.Get("typedb");
                Assert.AreEqual("typedb", database.Name);

                // Open transactions of 3 types
                var tx = driver.Transaction(database.Name, TransactionType.Read);

                // Use try blocks to catch driver exceptions
                try
                {
                    // Execute any TypeDB query using TypeQL. Wrong queries are rejected with an explicit exception
                    var answer = tx.Query("define entity i-cannot-be-defined-in-read-transactions;").Resolve()!;

                    Console.WriteLine("The query has been sent, iterating or committing will surface errors!");
                    // Iterating the answer would throw an exception
                }
                catch (TypeDBDriverException expectedException)
                {
                    Console.WriteLine("Query execution revealed the exception: " + expectedException);
                }
                finally
                {
                    // Don't forget to close the transaction!
                    tx.Close();
                }

                // Open a schema transaction to make schema changes
                // Transactions can be opened with configurable options. This option limits its lifetime
                var transactionOptions = new TransactionOptions { TransactionTimeoutMillis = 10_000 };

                // Use using blocks to forget about close operations (similarly to connections)
                using (var transaction = driver.Transaction(database.Name, TransactionType.Schema, transactionOptions))
                {
                    var defineQuery = @"define
                        entity person, owns name, owns age;
                        attribute name, value string;
                        attribute age, value integer;";

                    var answer = transaction.Query(defineQuery).Resolve()!;
                    Assert.IsTrue(answer.IsOk);
                    Assert.AreEqual(QueryType.Schema, answer.QueryType);

                    // Commit automatically closes the transaction. It can still be safely called inside using blocks
                    transaction.Commit();
                }

                // Open a read transaction to safely read anything without database modifications
                using (var transaction = driver.Transaction(database.Name, TransactionType.Read))
                {
                    var entityAnswer = transaction.Query("match entity $x;").Resolve()!;
                    Assert.IsTrue(entityAnswer.IsConceptRows);
                    Assert.IsFalse(entityAnswer.IsConceptDocuments);
                    Assert.AreEqual(QueryType.Read, entityAnswer.QueryType);

                    // Collect concept rows that represent the answer as a table
                    var entityRows = entityAnswer.AsConceptRows().ToList();
                    Assert.AreEqual(1, entityRows.Count);
                    var entityRow = entityRows[0];

                    // Collect column names to get concepts by index if the variable names are lost
                    var entityHeader = entityRow.ColumnNames.ToList();
                    Assert.AreEqual(1, entityHeader.Count);

                    var columnName = entityHeader[0];
                    Assert.AreEqual("x", columnName);

                    // Get concept by the variable name (column name)
                    var conceptByName = entityRow.Get(columnName);

                    // Get concept by the header's index
                    var conceptByIndex = entityRow.GetIndex(0);
                    Assert.AreEqual(conceptByName, conceptByIndex);

                    Console.WriteLine($"Getting concepts by variable names ({conceptByName?.TryGetLabel()}) and indexes ({conceptByIndex?.TryGetLabel()}) is equally correct. ");

                    // Check if it's an entity type before the conversion
                    if (conceptByName != null && conceptByName.IsEntityType())
                    {
                        Console.WriteLine($"Both represent the defined entity type: '{conceptByName.AsEntityType().GetLabel()}' (in case of a doubt: '{conceptByIndex?.AsEntityType().GetLabel()}')");
                    }
                    Assert.IsNotNull(conceptByName);
                    Assert.IsTrue(conceptByName!.IsEntityType());
                    Assert.IsTrue(conceptByName.IsType());
                    Assert.AreEqual("person", conceptByName.GetLabel());
                    Assert.AreEqual("person", conceptByName.AsEntityType().GetLabel());
                    Assert.AreNotEqual("not person", conceptByName.AsEntityType().GetLabel());
                    Assert.AreNotEqual("age", conceptByName.AsEntityType().GetLabel());

                    // Continue querying in the same transaction if needed
                    var attributeAnswer = transaction.Query("match attribute $a;").Resolve()!;
                    Assert.IsTrue(attributeAnswer.IsConceptRows);
                    Assert.AreEqual(QueryType.Read, attributeAnswer.QueryType);

                    // IConceptRowIterator can be used as any other IEnumerable
                    foreach (var attributeRow in attributeAnswer.AsConceptRows())
                    {
                        // Column names are an IEnumerable, so they can be used in a similar way
                        var columnNames = attributeRow.ColumnNames.ToList();
                        columnName = columnNames[0];
                        Assert.AreEqual(1, columnNames.Count);

                        conceptByName = attributeRow.Get(columnName);

                        // Check if it's an attribute type before the conversion
                        if (conceptByName != null && conceptByName.IsAttributeType())
                        {
                            var attributeType = conceptByName.AsAttributeType();
                            Console.WriteLine($"Defined attribute type's label: '{attributeType.GetLabel()}', value type: '{attributeType.TryGetValueType()}'");
                            Assert.IsTrue(attributeType.IsInteger() || attributeType.IsString());
                            Assert.IsTrue(attributeType.TryGetValueType() == "integer" || attributeType.TryGetValueType() == "string");
                            Assert.IsTrue(attributeType.GetLabel() == "age" || attributeType.GetLabel() == "name");
                            Assert.AreNotEqual("person", attributeType.GetLabel());
                            Assert.AreNotEqual("person:age", attributeType.GetLabel());

                            Console.WriteLine($"It is also possible to just print the concept itself: '{conceptByName}'");
                            Assert.IsTrue(conceptByName.IsAttributeType());
                            Assert.IsTrue(conceptByName.IsType());
                        }
                    }
                }

                // Open a write transaction to insert data
                using (var transaction = driver.Transaction(database.Name, TransactionType.Write))
                {
                    var insertQuery = "insert $z isa person, has age 10; $x isa person, has age 20, has name \"John\";";
                    var answer = transaction.Query(insertQuery).Resolve()!;
                    Assert.IsTrue(answer.IsConceptRows);
                    Assert.AreEqual(QueryType.Write, answer.QueryType);

                    // Insert queries also return concept rows
                    var rows = answer.AsConceptRows().ToList();
                    Assert.AreEqual(1, rows.Count);
                    var row = rows[0];
                    foreach (var colName in row.ColumnNames)
                    {
                        var insertedConcept = row.Get(colName);
                        Console.WriteLine($"Successfully inserted ${colName}: {insertedConcept}");
                        if (insertedConcept != null && insertedConcept.IsEntity())
                        {
                            Console.WriteLine("This time, it's an entity, not a type!");
                        }
                    }

                    // It is possible to ask for the column names again
                    var header = row.ColumnNames.ToList();
                    Assert.AreEqual(2, header.Count);
                    Assert.That(header, Does.Contain("x"));
                    Assert.That(header, Does.Contain("z"));

                    var x = row.GetIndex(header.IndexOf("x"));
                    Console.WriteLine($"As we expect an entity instance, we can try to get its IID (unique identification): {x?.TryGetIID()}. ");
                    if (x != null && x.IsEntity())
                    {
                        Console.WriteLine("It can also be retrieved directly and safely after a cast: " + x.AsEntity().TryGetIID());
                    }

                    // Do not forget to commit if the changes should be persisted
                    Console.WriteLine("CAUTION: Committing or closing (including leaving the using block) a transaction will invalidate all its uncollected answer iterators");
                    transaction.Commit();
                }

                // Open another write transaction to try inserting even more data
                using (var transaction = driver.Transaction(database.Name, TransactionType.Write))
                {
                    // When loading a large dataset, it's often better to batch queries.
                    // Just call commit, which will wait for all ongoing operations to finish before executing.
                    var queries = new[] { "insert $a isa person, has name \"Alice\";", "insert $b isa person, has name \"Bob\";" };
                    foreach (var query in queries)
                    {
                        transaction.Query(query).Resolve();
                    }
                    transaction.Commit();
                }

                using (var transaction = driver.Transaction(database.Name, TransactionType.Write))
                {
                    // In C#, query errors surface during query execution
                    try
                    {
                        var invalidQuery = "insert $c isa not-person, has name \"Chris\";";
                        transaction.Query(invalidQuery).Resolve();
                        Assert.Fail("TypeDBDriverException is expected");
                    }
                    catch (TypeDBDriverException expectedException)
                    {
                        Console.WriteLine("Query execution revealed the error: " + expectedException);
                    }
                    // Transaction is still usable after catching the exception
                }

                // Open a read transaction to verify that the inserted data is saved
                using (var transaction = driver.Transaction(database.Name, TransactionType.Read))
                {
                    // Queries can also be executed with configurable options. This option forces the database
                    // to include types of instance concepts in ConceptRows answers
                    var queryOptions = new QueryOptions { IncludeInstanceTypes = true };
                    // A match query can be used for concept row outputs
                    var varName = "x";
                    var matchAnswer = transaction.Query($"match ${varName} isa person;", queryOptions).Resolve()!;
                    Assert.IsTrue(matchAnswer.IsConceptRows);
                    Assert.AreEqual(QueryType.Read, matchAnswer.QueryType);

                    // Simple match queries always return concept rows
                    var matchCount = 0;
                    foreach (var row in matchAnswer.AsConceptRows())
                    {
                        Assert.AreEqual(QueryType.Read, row.QueryType);
                        var x = row.Get(varName);
                        Assert.IsNotNull(x);
                        Assert.IsTrue(x!.IsEntity());
                        Assert.IsFalse(x.IsEntityType());
                        Assert.IsFalse(x.IsAttribute());
                        Assert.IsFalse(x.IsType());
                        Assert.IsTrue(x.IsInstance());
                        var xType = x.AsEntity().Type.AsEntityType();
                        Assert.AreEqual("person", xType.GetLabel());
                        Assert.AreNotEqual("not person", xType.GetLabel());
                        matchCount++;
                        Console.WriteLine($"Found a person {x} of type {xType}");
                    }
                    Assert.AreEqual(4, matchCount);
                    Console.WriteLine("Total persons found: " + matchCount);

                    // A fetch query can be used for concept document outputs with flexible structure
                    var fetchAnswer = transaction.Query(@"match
                        $x isa! person, has $a;
                        $a isa! $t;
                        fetch {
                            ""single attribute type"": $t,
                            ""single attribute"": $a,
                            ""all attributes"": { $x.* },
                        };").Resolve()!;
                    Assert.IsTrue(fetchAnswer.IsConceptDocuments);
                    Assert.AreEqual(QueryType.Read, fetchAnswer.QueryType);

                    // Fetch queries always return concept documents
                    var fetchCount = 0;
                    foreach (var document in fetchAnswer.AsConceptDocuments())
                    {
                        Assert.IsNotNull(document);
                        Console.WriteLine("Fetched a document: " + document);
                        Console.Write("This document contains an attribute of type: ");
                        Console.WriteLine(document.AsObject()["single attribute type"].AsObject()["label"]);

                        fetchCount++;
                    }
                    Assert.AreEqual(5, fetchCount);
                    Console.WriteLine("Total documents fetched: " + fetchCount);
                }
            }

            Console.WriteLine("More examples can be found in the API reference and the documentation.\nWelcome to TypeDB!");
        }
    }
}
// EXAMPLE END MARKER
